Changing `monitor_type` in-place corrupts TF state and panics on refresh

## Summary

The API rejects changing `monitor_type` from `keyword` to `playwright` in place - the [`PATCH` request](https://betterstack.com/docs/uptime/api/update-an-existing-monitor/) returns 422. Due to [hashicorp/terraform-plugin-sdk#476](https://github.com/hashicorp/terraform-plugin-sdk/issues/476), Terraform still persists the planned (invalid) attributes into state. The next `plan` then panics while refreshing that resource (nil dereference).

## Steps to reproduce

1. Deploy a `keyword` monitor
```terraform
resource "betteruptime_monitor" "httpbingo_probe" {
  check_frequency    = 300
  http_method        = "get"
  monitor_type       = "keyword"
  policy_id          = data.betteruptime_policy.xyz.id
  pronounceable_name = "httpbingo probe"
  push               = false
  regions            = ["us"]
  required_keyword   = "slideshow"
  request_timeout    = 30
  url                = "https://httpbingo.org/json"
}
```

2. The following data was persistent in the Terraform state
```console
$ terragrunt state show 'betteruptime_monitor.httpbingo_probe'
# betteruptime_monitor.httpbingo_probe:
resource "betteruptime_monitor" "httpbingo_probe" {
    auth_password         = (sensitive value)
    auth_username         = (sensitive value)
    call                  = false
    check_frequency       = 300
    confirmation_period   = 0
    created_at            = "2026-07-23T15:14:21.233Z"
    critical_alert        = false
    domain_expiration     = -1
    email                 = true
    environment_variables = (sensitive value)
    expiration_policy_id  = 0
    follow_redirects      = true
    http_method           = "get"
    id                    = "REDACTED"
    ip_version            = null
    last_checked_at       = null
    maintenance_days      = [
        "mon",
        "tue",
        "wed",
        "thu",
        "fri",
        "sat",
        "sun",
    ]
    maintenance_from      = null
    maintenance_timezone  = null
    maintenance_to        = null
    monitor_group_id      = 0
    monitor_type          = "keyword"
    paused                = false
    paused_at             = null
    playwright_script     = null
    policy_id             = null
    port                  = null
    pronounceable_name    = "httpbingo probe"
    proxy_host            = null
    proxy_port            = 0
    push                  = false
    recovery_period       = 180
    regions               = [
        "us",
    ]
    remember_cookies      = true
    request_body          = null
    request_timeout       = 30
    required_keyword      = "slideshow"
    sms                   = false
    ssl_expiration        = -1
    status                = "pending"
    team_wait             = 0
    updated_at            = "2026-07-23T15:14:21.233Z"
    url                   = "https://httpbingo.org/json"
    verify_ssl            = true
}
```

3. Change the monitor type to `playwright` ([docs](https://betterstack.com/docs/uptime/playwright-monitor/)) and run `apply`
```terraform
resource "betteruptime_monitor" "httpbingo_probe" {
  check_frequency    = 300
  monitor_type       = "playwright"
  policy_id          = data.betteruptime_policy.xyz.id
  pronounceable_name = "httpbingo probe"
  push               = false
  regions            = ["us"]
  request_timeout    = 30
  scenario_name      = "httpbingo probe"
  playwright_script  = <<-EOT
    import { test, expect } from '@playwright/test';

    test('httpbingo returns slideshow JSON', async () => {
      const response = await fetch('https://httpbingo.org/json');
      expect(response.ok).toBeTruthy();
      const data = await response.json();
      expect(data.slideshow).toBeTruthy();
    });
  EOT
}
```

4. `terraform apply` ends with the following error:
```console
╷
│ Error: PATCH https://uptime.betterstack.com/api/v2/monitors/REDACTED returned 422: {"errors":{"monitor_type":["is invalid. Valid values are: status, expected_status_code, keyword, keyword_absence"]}}
│

---

2026-07-23T17:35:37.984+0200 [INFO]  provider.terraform-provider-better-uptime_v0.21.10: resource.go:77: PATCH /api/v2/monitors/REDACTED: {"expiration_policy_id":null,"url":"","monitor_type":"playwright","required_keyword":"","playwright_script":"import { test, expect } from '@playwright/test';\n\ntest('httpbingo returns slideshow JSON', async () =\u003e {\n  const response = await fetch('https://httpbingo.org/json');\n  expect(response.ok).toBeTruthy();\n  const data = await response.json();\n  expect(data.slideshow).toBeTruthy();\n});\n","scenario_name":"httpbingo probe"}: timestamp="2026-07-23T17:35:37.984+0200"
2026-07-23T17:35:38.141+0200 [ERROR] provider.terraform-provider-better-uptime_v0.21.10: Response contains error diagnostic: diagnostic_detail="" diagnostic_severity=ERROR diagnostic_summary="PATCH https://uptime.betterstack.com/api/v2/monitors/REDACTED returned 422: {\"errors\":{\"monitor_type\":[\"is invalid. Valid values are: status, expected_status_code, keyword, keyword_absence\"]}}" tf_proto_version=5.8 tf_req_id=68e947b1-deb4-c259-41c7-40d6f650095c @caller=github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov5/internal/diag/diagnostics.go:58 @module=sdk.proto tf_provider_addr=provider tf_resource_type=betteruptime_monitor tf_rpc=ApplyResourceChange timestamp="2026-07-23T17:35:38.141+0200"
```

5. The state contains the following resource state now:
```console
$ terragrunt state show 'betteruptime_monitor.httpbingo_probe'
# betteruptime_monitor.httpbingo_probe:
resource "betteruptime_monitor" "httpbingo_probe" {
    auth_password         = (sensitive value)
    auth_username         = (sensitive value)
    call                  = false
    check_frequency       = 300
    confirmation_period   = 0
    created_at            = "2026-07-23T15:14:21.233Z"
    critical_alert        = false
    domain_expiration     = -1
    email                 = true
    environment_variables = (sensitive value)
    expected_status_codes = []
    expiration_policy_id  = 0
    follow_redirects      = true
    http_method           = "get"
    id                    = "REDACTED"
    ip_version            = null
    last_checked_at       = "2026-07-23T15:34:02.000Z"
    maintenance_days      = [
        "mon",
        "tue",
        "wed",
        "thu",
        "fri",
        "sat",
        "sun",
    ]
    maintenance_from      = null
    maintenance_timezone  = null
    maintenance_to        = null
    monitor_group_id      = 0
    monitor_type          = "playwright"
    paused                = false
    paused_at             = null
    playwright_script     = <<-EOT
        import { test, expect } from '@playwright/test';

        test('httpbingo returns slideshow JSON', async () => {
          const response = await fetch('https://httpbingo.org/json');
          expect(response.ok).toBeTruthy();
          const data = await response.json();
          expect(data.slideshow).toBeTruthy();
        });
    EOT
    policy_id             = null
    port                  = null
    pronounceable_name    = "httpbingo probe"
    proxy_host            = null
    proxy_port            = 0
    push                  = false
    recovery_period       = 180
    regions               = [
        "us",
    ]
    remember_cookies      = true
    request_body          = null
    request_headers       = []
    request_timeout       = 30
    required_keyword      = null
    scenario_name         = "httpbingo probe"
    sms                   = false
    ssl_expiration        = -1
    status                = "up"
    team_wait             = 0
    updated_at            = "2026-07-23T15:14:21.233Z"
    url                   = null
    verify_ssl            = true
}
```

6. The next `terraform plan` result in a panic
```console
$ terraform plan
...
Stack trace from the terraform-provider-better-uptime_v0.21.10 plugin:

  panic: runtime error: invalid memory address or nil pointer dereference
  [signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1007d1f54]

  goroutine 550 [running]:
  github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider.monitorCopyAttrs(0x140005a8480, 0x14000766350)
        github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider/resource_monitor.go:628 +0x314
  github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider.monitorRead({0x100d1e628, 0x14000722700}, 0x140005a8480, {0x100cba900, 0x1400029f180})
        github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider/resource_monitor.go:608 +0x150
  github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0x14000335900, {0x100d1e580, 0x14000816000}, 0x140005a8480, {0x100cba900, 0x1400029f180})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.36.1/helper/schema/resource.go:855 +0xe4
  github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0x14000335900, {0x100d1e580, 0x14000816000}, 0x140002c0a90, {0x100cba900, 0x1400029f180})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.36.1/helper/schema/resource.go:1149 +0x404
  github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0x140004145a0, {0x100d1e580?, 0x1400012bec0?}, 0x14000422480)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.36.1/helper/schema/grpc_provider.go:739 +0x5b4
  github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0x14000307a40, {0x100d1e580?, 0x1400012ab70?}, 0x14000722070)
        github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov5/tf5server/server.go:784 +0x21c
  github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0x100cf1600, 0x14000307a40}, {0x100d1e580, 0x1400012ab70}, 0x140005a8000, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.26.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:575 +0x1c0
  google.golang.org/grpc.(*Server).processUnaryRPC(0x14000304200, {0x100d1e580, 0x1400012a720}, 0x140001ec600, 0x14000426330, 0x1012ac888, 0x0)
        google.golang.org/grpc@v1.69.4/server.go:1392 +0xc44
  google.golang.org/grpc.(*Server).handleStream(0x14000304200, {0x100d1ecc8, 0x1400033b860}, 0x140001ec600)
        google.golang.org/grpc@v1.69.4/server.go:1802 +0x910
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.69.4/server.go:1030 +0x84
  created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 26
        google.golang.org/grpc@v1.69.4/server.go:1041 +0x13c

  Error: The terraform-provider-better-uptime_v0.21.10 plugin crashed!

  This is always indicative of a bug within the plugin. It would be immensely
  helpful if you could report the crash with the plugin's maintainers so that it
  can be fixed. The output above should help diagnose the issue.
```

## Current state

1. Corrupted attributes remain in Terraform state even though the API rejected the update and the remote monitor is unchanged
2. Subsequent `plan` crashes the provider

## Expected state

1. Changing `monitor_type` should force resource replacement, not an in-place update the API cannot apply
2. The provider should nil-check optional fields instead of assuming they are always present

## Noteworthy information

- Provider version: `0.21.10` (latest at time of writing)
- Related Terraform Plugin SDK issue: https://github.com/hashicorp/terraform-plugin-sdk/issues/476
- Recovery requires a manual state fix or `terraform state rm <resource>` combined with a re-import

--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,7 @@ GOLANGCI_LINT := golangci-lint run --disable-all \
 	-E staticcheck \
 	-E typecheck \
 	-E unused
-VERSION := 0.21.10
+VERSION := 0.21.11
 .PHONY: test build
 
 help:

--- a/internal/provider/resource_monitor.go
+++ b/internal/provider/resource_monitor.go
@@ -620,16 +620,17 @@ func monitorCopyAttrs(d *schema.ResourceData, in *monitor) diag.Diagnostics {
 				derr = append(derr, diag.FromErr(err)[0])
 			}
 		} else if e.k == "url" {
-			// Special handling for URL and scenario name
+			// Special handling for URL and scenario name. The API omits either field for
+			// monitor types that don't have it, so keep the state value when it's absent.
 			currentUrl := d.Get("url").(string)
 			currentScenarioName := d.Get("scenario_name").(string)
-			if currentScenarioName != "" {
+			if currentScenarioName != "" && in.ScenarioName != nil {
 				// Read scenario name from API only if we have it defined
 				if err := d.Set("scenario_name", *in.ScenarioName); err != nil {
 					derr = append(derr, diag.FromErr(err)[0])
 				}
 			}
-			if currentUrl != "" || currentScenarioName == "" {
+			if (currentUrl != "" || currentScenarioName == "") && in.URL != nil {
 				// Read URL from API if we have it defined, or if we're missing scenario name
 				if err := d.Set("url", *in.URL); err != nil {
 					derr = append(derr, diag.FromErr(err)[0])
@@ -698,6 +699,17 @@ func validateMonitor(ctx context.Context, diff *schema.ResourceDiff, v interface
 	// Validate request headers
 	if err := validateRequestHeaders(ctx, diff, v); err != nil {
 		return err
+	}
+
+	// The API rejects in-place monitor_type changes to or from playwright (422), and the
+	// failed apply would still persist the planned attributes into state
+	// (hashicorp/terraform-plugin-sdk#476), leaving a state that panics on refresh - fail
+	// at plan time instead. Changes among the other types stay allowed in place.
+	if diff.Id() != "" && diff.HasChange("monitor_type") {
+		oldType, newType := diff.GetChange("monitor_type")
+		if oldType.(string) == "playwright" || newType.(string) == "playwright" {
+			return fmt.Errorf("monitor_type cannot be changed to or from 'playwright' in place; recreate the monitor instead, e.g. via terraform apply -replace")
+		}
 	}
 
 	// Validate URL requirement based on monitor type

--- a/internal/provider/resource_monitor_type_change_test.go
+++ b/internal/provider/resource_monitor_type_change_test.go
@@ -1,0 +1,155 @@
+package provider
+
+import (
+	"regexp"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// The API omits scenario_name and url for monitor types that don't have them;
+// monitorCopyAttrs must not dereference the missing fields (issue #228: a refresh
+// after a failed in-place type change panicked with a nil dereference).
+func TestMonitorCopyAttrsNilOptionalFields(t *testing.T) {
+	t.Run("scenario_name in state, absent in API response", func(t *testing.T) {
+		d := schema.TestResourceDataRaw(t, monitorSchema, map[string]interface{}{
+			"scenario_name": "httpbingo probe",
+			"monitor_type":  "keyword",
+		})
+		monitorType := "keyword"
+		monitorUrl := "https://example.com"
+		if derr := monitorCopyAttrs(d, &monitor{MonitorType: &monitorType, URL: &monitorUrl}); derr != nil {
+			t.Fatal(derr)
+		}
+		if got := d.Get("scenario_name").(string); got != "httpbingo probe" {
+			t.Fatalf("scenario_name = %q, want the state value preserved", got)
+		}
+	})
+
+	t.Run("url in state, absent in API response", func(t *testing.T) {
+		d := schema.TestResourceDataRaw(t, monitorSchema, map[string]interface{}{
+			"url":          "https://example.com",
+			"monitor_type": "playwright",
+		})
+		monitorType := "playwright"
+		if derr := monitorCopyAttrs(d, &monitor{MonitorType: &monitorType}); derr != nil {
+			t.Fatal(derr)
+		}
+		if got := d.Get("url").(string); got != "https://example.com" {
+			t.Fatalf("url = %q, want the state value preserved", got)
+		}
+	})
+}
+
+func TestResourceMonitorTypeChangeToPlaywright(t *testing.T) {
+	server := createTestServer(t)
+	defer server.Close()
+
+	resource.Test(t, resource.TestCase{
+		IsUnitTest: true,
+		ProviderFactories: map[string]func() (*schema.Provider, error){
+			"betteruptime": func() (*schema.Provider, error) {
+				return New(WithURL(server.URL)), nil
+			},
+		},
+		Steps: []resource.TestStep{
+			// Step 1 - create a keyword monitor.
+			{
+				Config: `
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					url              = "https://example.com"
+					monitor_type     = "keyword"
+					required_keyword = "example"
+				}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("betteruptime_monitor.this", "monitor_type", "keyword"),
+				),
+			},
+			// Step 2 - in-place changes within the non-playwright types stay allowed.
+			{
+				Config: `
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					url          = "https://example.com"
+					monitor_type = "status"
+				}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("betteruptime_monitor.this", "monitor_type", "status"),
+				),
+			},
+			// Step 3 - changing to playwright must fail at plan time: the API rejects it
+			// and the failed apply would corrupt state (issue #228).
+			{
+				Config: `
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					monitor_type      = "playwright"
+					scenario_name     = "example probe"
+					playwright_script = "// script"
+				}
+				`,
+				ExpectError: regexp.MustCompile(`monitor_type cannot be changed to or from 'playwright'`),
+			},
+		},
+	})
+}
+
+func TestResourceMonitorTypeChangeFromPlaywright(t *testing.T) {
+	server := createTestServer(t)
+	defer server.Close()
+
+	resource.Test(t, resource.TestCase{
+		IsUnitTest: true,
+		ProviderFactories: map[string]func() (*schema.Provider, error){
+			"betteruptime": func() (*schema.Provider, error) {
+				return New(WithURL(server.URL)), nil
+			},
+		},
+		Steps: []resource.TestStep{
+			// Step 1 - create a playwright monitor.
+			{
+				Config: `
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					monitor_type      = "playwright"
+					scenario_name     = "example probe"
+					playwright_script = "// script"
+				}
+				`,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("betteruptime_monitor.this", "monitor_type", "playwright"),
+				),
+			},
+			// Step 2 - changing away from playwright must fail at plan time.
+			{
+				Config: `
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					url          = "https://example.com"
+					monitor_type = "status"
+				}
+				`,
+				ExpectError: regexp.MustCompile(`monitor_type cannot be changed to or from 'playwright'`),
+			},
+		},
+	})
+}

